### PR TITLE
Handle row-version columns in owned types with conversions

### LINKED ISSUE #29689: Entity Configuration using IsRowVersion.HasConversion<byte[]> for Sql Server fails if derived entity uses OwnsOne config
I became interested in the conversion of the SQL Server row version from a `byte[]` into a `ulong` after reading [the docs here](https://learn.microsoft.com/en-us/ef/core/modeling/value-conversions?tabs=data-annotations#use-ulong-for-timestamprowversion) 

Everything worked great except for a TPH that had a base class `NotificationChannelBase` and three derived types:
- `EmailChannel`
- `PushChannel`
- `SmsChannel`

The base class contains this:
```csharp
public ulong Version { get; init; }
```

and the configuration class contains this:
```csharp
builder
  .Property(e => e.Version)
  .IsRowVersion()
  .HasConversion<byte[]>();
```

An error occurs when doing a save changes - _but only for PushChannel and SmsChannel as they contain value objects as owned properties._

The error is: `Unable to cast object of type 'System.UInt64' to type 'System.Byte[]'`

The `EmailChannel` works fine.  I simulated what was happening with another class to confirm it.  `OwnsOne` seems to throw it off.

I subsequently modified `NotificationChannelBase` to use a `byte[] Version`, removed the conversion inside the configuration and it's all back to working.

This approach worked for all my other objects.  It would be great if I could use a single approach within my entities.

HTH

Only the PushChannel and SMS Channel have owned types,
EF Core version:  7.0.0
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET 7.0
Operating system:  Windows 11 (latest)
IDE: Visual Studio 2022 17.4.0 Preview 6.0

### PR DESCRIPTION
**Pull request by ajcvickers**
Fixes #29689

Also add more end-to-end tests for optimistic concurrency, converters, owned types, inheritance, and table sharing.

### CODE REVIEW

File: test/EFCore.SqlServer.FunctionalTests/OptimisticConcurrencySqlServerTest.cs [hunk 1]
```diff
@@ -51,6 +123,179 @@ protected OptimisticConcurrencySqlServerTestBase(TFixture fixture)
...
+                if (updateDependent)
+                {
+                    fanInner.City.Name += "-";
+                }
+                else
+                {
+                    fanInner.Name += "-";
+                }
+
+                if (mapping == Mapping.Tpc)
```
Comment (ajcvickers) [inline test/EFCore.SqlServer.FunctionalTests/OptimisticConcurrencySqlServerTest.cs:263]: @AndriySvyryd For TPC with table splitting, this throws:

```
System.InvalidOperationException: The property 'CityTpc.Name' cannot be assigned a value generated by the database. Store-generated values can only be...

System.InvalidOperationException
The property 'CityTpc.Name' cannot be assigned a value generated by the database. Store-generated values can only be assigned to properties configured to use store-generated values.
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.InternalEntityEntry.SetStoreGeneratedValue(IProperty property, Object value, Boolean setModified) in C:\github\efcore\src\EFCore\ChangeTracking\Internal\InternalEntityEntry.cs:line 794
   at Microsoft.EntityFrameworkCore.Update.ModificationCommand.ColumnValuePropagator.TryPropagate(IColumnMappingBase mapping, IUpdateEntry entry) in C:\github\efcore\src\EFCore.Relational\Update\ModificationCommand.cs:line 984
   at Microsoft.EntityFrameworkCore.Update.ModificationCommand.<GenerateColumnModifications>g__HandleColumnModification|41_5(IColumnMappingBase columnMapping, <>c__DisplayClass41_0&, <>c__DisplayClass41_1&, <>c__DisplayClass41_2&) in C:\github\efcore\src\EFCore.Relational\Update\ModificationCommand.cs:line 550
   at Microsoft.EntityFrameworkCore.Update.ModificationCommand.GenerateColumnModifications() in C:\github\efcore\src\EFCore.Relational\Update\ModificationCommand.cs:line 407
   at Microsoft.EntityFrameworkCore.Update.ModificationCommand.<>c.<get_ColumnModifications>b__33_0(ModificationCommand command) in C:\github\efcore\src\EFCore.Relational\Update\ModificationCommand.cs:line 101
   at Microsoft.EntityFrameworkCore.Internal.NonCapturingLazyInitializer.EnsureInitialized[TParam,TValue](TValue& target, TParam param, Func`2 valueFactory) in C:\github\efcore\src\Shared\NonCapturingLazyInitializer.cs:line 27
   at Microsoft.EntityFrameworkCore.Update.ModificationCommand.get_ColumnModifications() in C:\github\efcore\src\EFCore.Relational\Update\ModificationCommand.cs:line 100
   at Microsoft.EntityFrameworkCore.Update.Internal.CommandBatchPreparer.CreateCommandBatches(IEnumerable`1 commandSet, Boolean moreCommandSets, Boolean assertColumnModification, ParameterNameGenerator parameterNameGenerator)+MoveNext() in C:\github\efcore\src\EFCore.Relational\Update\Internal\CommandBatchPreparer.cs:line 119
   at Microsoft.EntityFrameworkCore.Update.Internal.CommandBatchPreparer.BatchCommands(IList`1 entries, IUpdateAdapter updateAdapter)+MoveNext() in C:\github\efcore\src\EFCore.Relational\Update\Internal\CommandBatchPreparer.cs:line 78
   at Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.ExecuteAsync(IEnumerable`1 commandBatches, IRelationalConnection connection, CancellationToken cancellationToken) in C:\github\efcore\src\EFCore.Relational\Update\Internal\BatchExecutor.cs:line 165
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChangesAsync(IList`1 entriesToSave, CancellationToken cancellationToken) in C:\github\efcore\src\EFCore\ChangeTracking\Internal\StateManager.cs:line 1226
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChangesAsync(StateManager stateManager, Boolean acceptAllChangesOnSuccess, CancellationToken cancellationToken) in C:\github\efcore\src\EFCore\ChangeTracking\Internal\StateManager.cs:line 1325
   at Microsoft.EntityFrameworkCore.Storage.ExecutionStrategy.ExecuteAsync[TState,TResult](TState state, Func`4 operation, Func`4 verifySucceeded, CancellationToken cancellationToken) in C:\github\efcore\src\EFCore\Storage\ExecutionStrategy.cs:line 304
   at Microsoft.EntityFrameworkCore.DbContext.SaveChangesAsync(Boolean acceptAllChangesOnSuccess, CancellationToken cancellationToken) in C:\github\efcore\src\EFCore\DbContext.cs:line 771
   at Microsoft.EntityFrameworkCore.DbContext.SaveChangesAsync(Boolean acceptAllChangesOnSuccess, CancellationToken cancellationToken) in C:\github\efcore\src\EFCore\DbContext.cs:line 808
   at Microsoft.EntityFrameworkCore.OptimisticConcurrencySqlServerTestBase`2.<>c__DisplayClass3_0`3.<<Row_version_with_table_splitting>b__0>d.MoveNext() in C:\github\efcore\test\EFCore.SqlServer.FunctionalTests\OptimisticConcurrencySqlServerTest.cs:line 261
--- End of stack trace from previous location ---
   at Microsoft.EntityFrameworkCore.OptimisticConcurrencySqlServerTestBase`2.<>c__DisplayClass3_0`3.<<Row_version_with_table_splitting>b__0>d.MoveNext() in C:\github\efcore\test\EFCore.SqlServer.FunctionalTests\OptimisticConcurrencySqlServerTest.cs:line 285
--- End of stack trace from previous location ---
   at Microsoft.EntityFrameworkCore.OptimisticConcurrencySqlServerTestBase`2.<>c__DisplayClass3_0`3.<<Row_version_with_table_splitting>b__0>d.MoveNext() in C:\github\efcore\test\EFCore.SqlServer.FunctionalTests\OptimisticConcurrencySqlServerTest.cs:line 285
--- End of stack trace from previous location ---
   at Microsoft.EntityFrameworkCore.ExecutionStrategyExtensions.<>c__7`1.<<ExecuteAsync>b__7_0>d.MoveNext() in C:\github\efcore\src\EFCore\Storage\ExecutionStrategyExtensions.cs:line 238
--- End of stack trace from previous location ---
   at Microsoft.EntityFrameworkCore.Storage.ExecutionStrategy.<>c__DisplayClass30_0`2.<<ExecuteAsync>b__0>d.MoveNext() in C:\github\efcore\src\EFCore\Storage\ExecutionStrategy.cs:line 311
--- End of stack trace from previous location ---
   at Microsoft.EntityFrameworkCore.Storage.ExecutionStrategy.ExecuteImplementationAsync[TState,TResult](Func`4 operation, Func`4 verifySucceeded, TState state, CancellationToken cancellationToken) in C:\github\efcore\src\EFCore\Storage\ExecutionStrategy.cs:line 334
   at Microsoft.EntityFrameworkCore.Storage.ExecutionStrategy.ExecuteImplementationAsync[TState,TResult](Func`4 operation, Func`4 verifySucceeded, TState state, CancellationToken cancellationToken) in C:\github\efcore\src\EFCore\Storage\ExecutionStrategy.cs:line 358
   at Microsoft.EntityFrameworkCore.Storage.ExecutionStrategy.ExecuteAsync[TState,TResult](TState state, Func`4 operation, Func`4 verifySucceeded, CancellationToken cancellationToken) in C:\github\efcore\src\EFCore\Storage\ExecutionStrategy.cs:line 310
   at Microsoft.EntityFrameworkCore.OptimisticConcurrencySqlServerTestBase`2.Row_version_with_table_splitting[TEntity,TCity,TVersion](Boolean updateDependent, Mapping mapping, String propertyName) in C:\github\efcore\test\EFCore.SqlServer.FunctionalTests\OptimisticConcurrencySqlServerTest.cs:line 217
   at Microsoft.EntityFrameworkCore.OptimisticConcurrencySqlServerTestBase`2.Row_version_with_table_splitting[TEntity,TCity,TVersion](Boolean updateDependent, Mapping mapping, String propertyName) in C:\github\efcore\test\EFCore.SqlServer.FunctionalTests\OptimisticConcurrencySqlServerTest.cs:line 217
   at Xunit.Sdk.TestInvoker`1.<>c__DisplayClass48_0.<<InvokeTestMethodAsync>b__1>d.MoveNext() in /_/src/xunit.execution/Sdk/Frameworks/Runners/TestInvoker.cs:line 264
--- End of stack trace from previous location ---
   at Xunit.Sdk.ExecutionTimer.AggregateAsync(Func`1 asyncAction) in /_/src/xunit.execution/Sdk/Frameworks/ExecutionTimer.cs:line 48
   at Xunit.Sdk.ExceptionAggregator.RunAsync(Func`1 code) in /_/src/xunit.core/Sdk/ExceptionAggregator.cs:line 90
```

File: test/EFCore.SqlServer.FunctionalTests/OptimisticConcurrencySqlServerTest.cs [hunk 2]
```diff
@@ -51,6 +123,179 @@ protected OptimisticConcurrencySqlServerTestBase(TFixture fixture)
...
+                    if (mapping == Mapping.Tph)
+                    {
+                        var swagVersion = swagEntry.Property<TVersion>(synthesizedPropertyName).CurrentValue;
+                        Assert.Equal(fanVersion, swagVersion);
+                        Assert.NotEqual(originalSwagVersion, swagVersion);
+                    }
+                }
+                else
+                {
+                    await context.SaveChangesAsync();
```
Comment (ajcvickers) [inline test/EFCore.SqlServer.FunctionalTests/OptimisticConcurrencySqlServerTest.cs:205]: For TPT when the owned entity is updated, there is no `DbUpdateConcurrencyException`

File: test/EFCore.SqlServer.FunctionalTests/OptimisticConcurrencySqlServerTest.cs [hunk 3]
```diff
@@ -51,6 +123,179 @@ protected OptimisticConcurrencySqlServerTestBase(TFixture fixture)
...
+
+    protected async Task Row_version_with_owned_types<TEntity, TVersion>(bool updateOwned, Mapping mapping, string propertyName)
+        where TEntity : class, ISuperFan
+    {
+        await using var c = CreateF1Context();
+
+        await c.Database.CreateExecutionStrategy().ExecuteAsync(
+            c, async context =>
+            {
+                var synthesizedPropertyName = $"_TableSharingConcurrencyTokenConvention_{propertyName}";
```
Comment (ajcvickers) [inline test/EFCore.SqlServer.FunctionalTests/OptimisticConcurrencySqlServerTest.cs:141]: @AndriySvyryd This property is only present in TPH cases.

**Comment by ajcvickers** [comment]
@AndriySvyryd I have made three comments which might indicate additional bugs, either known or nor.

File: src/EFCore.Relational/Metadata/Conventions/TableSharingConcurrencyTokenConvention.cs [hunk 4]
```diff
@@ -103,17 +103,38 @@ public class TableSharingConcurrencyTokenConvention : IModelFinalizingConvention
...
-                        .ValueGenerated(exampleProperty.ValueGenerated);
+                        .ValueGenerated(exampleProperty.ValueGenerated)!;
+
+                    var typeMapping = exampleProperty.FindTypeMapping();
+                    if (typeMapping != null)
+                    {
+                        propertyBuilder = propertyBuilder.HasTypeMapping(typeMapping)!;
+                    }
+
+                    var converter = exampleProperty.GetValueConverter();
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Metadata/Conventions/TableSharingConcurrencyTokenConvention.cs:121]: Use `GetValueConverterType` if it's not `null` instead

File: test/EFCore.SqlServer.FunctionalTests/OptimisticConcurrencySqlServerTest.cs (hunk 1 above)
Comment (AndriySvyryd) [reply test/EFCore.SqlServer.FunctionalTests/OptimisticConcurrencySqlServerTest.cs:263]: It should be `ValueGenerated.OnUpdateSometimes`

**Review by AndriySvyryd — APPROVED** [review]
Add another update to the tests and file the 2 new issues

**Pull request merged**